src/core/ext/filters/client_channel/lb_policy/weighted_target/weighted_target.cc

```diff
@@ -401,6 +413,53 @@ void WeightedTargetLb::UpdateStateLocked() {
                                         std::move(picker));
 }
 
+//
+// WeightedTargetLb::WeightedChild::DelayedRemovalTimer
+//
+
+WeightedTargetLb::WeightedChild::DelayedRemovalTimer::DelayedRemovalTimer(
+    RefCountedPtr<WeightedTargetLb::WeightedChild> weighted_child)
+    : weighted_child_(std::move(weighted_child)) {
+  GRPC_CLOSURE_INIT(&on_timer_, OnTimer, this, nullptr);
+  Ref().release();
+  grpc_timer_init(&timer_, ExecCtx::Get()->Now() + kChildRetentionInterval,
+                  &on_timer_);
+}
+
+void WeightedTargetLb::WeightedChild::DelayedRemovalTimer::Orphan() {
+  if (timer_pending_) {
+    if (GRPC_TRACE_FLAG_ENABLED(grpc_lb_weighted_target_trace)) {
+      gpr_log(GPR_INFO,
+              "[weighted_target_lb %p] WeightedChild %p %s: cancelling "
+              "delayed removal timer",
+              weighted_child_->weighted_target_policy_.get(),
+              weighted_child_.get(), weighted_child_->name_.c_str());
+    }
+    timer_pending_ = false;
+    grpc_timer_cancel(&timer_);
+  }
+}
+
+void WeightedTargetLb::WeightedChild::DelayedRemovalTimer::OnTimer(
+    void* arg, grpc_error_handle error) {
+  auto* self = static_cast<DelayedRemovalTimer*>(arg);
+  (void)GRPC_ERROR_REF(error);  // ref owned by lambda
+  self->weighted_child_->weighted_target_policy_->work_serializer()->Run(
+      [self, error]() { self->OnTimerLocked(error); }, DEBUG_LOCATION);
+}
+
+void WeightedTargetLb::WeightedChild::DelayedRemovalTimer::OnTimerLocked(
+    grpc_error_handle error) {
+  if (error == GRPC_ERROR_NONE && timer_pending_ &&
+      !weighted_child_->shutdown_ && weighted_child_->weight_ == 0) {
+    timer_pending_ = false;
```

Comment:
nit: any reason not to set `timer_pending_ = false`, unconditionally from this method?

Otherwise, I think there would be no bug, but `Orphan` would still log `cancelling delayed removal timer`, which might be confusing.

Same suggestion for timers in priority lb PR too, actually: https://github.com/grpc/grpc/pull/29188/files

Reply:
I don't think that case is actually possible, because in basically any of those other cases, `timer_pending_` will already be false.

I think this means that we actually don't need to check `shutdown_` or `weight_` at all, so I've changed the code accordingly.